packages/lesswrong/components/hooks/useExpandedFrontpageSection.tsx

```diff
@@ -0,0 +1,101 @@
+import { useState, useCallback } from "react";
+import { useCurrentUser } from "../common/withUser";
+import { useTracking } from "../../lib/analyticsEvents";
+import { useCookiesWithConsent } from "./useCookiesWithConsent";
+import { useMutation, gql } from "@apollo/client";
+import moment from "moment";
+
+export type ExpandedFrontpageSections = NonNullable<DbUser["expandedFrontpageSections"]>;
+export type ExpandedFrontpageSection = keyof ExpandedFrontpageSections;
+export type DefaultExpandedType = "all" | "none" | "loggedIn";
+
+export type UseExpandedFrontpageSectionProps = {
+  section: ExpandedFrontpageSection,
+  defaultExpanded: DefaultExpandedType,
+  onExpandEvent?: string,
+  onCollapseEvent?: string,
+  cookieName?: string,
+}
+
+const expandFrontpageSectionMutation = gql`
+  mutation UserExpandFrontpageSection($section: String!, $expanded: Boolean!) {
+    UserExpandFrontpageSection(section: $section, expanded: $expanded)
+  }
+`;
+
+const isDefaultExpanded = (
+  currentUser: UsersCurrent | null,
+  defaultExpanded: DefaultExpandedType,
+): boolean => {
+  switch (defaultExpanded) {
+  case "none":
+      return false;
+  case "all":
+      return true;
+  case "loggedIn":
+      return !!currentUser;
+  }
+}
+
+export const useExpandedFrontpageSection = ({
+  section,
+  defaultExpanded,
+  onExpandEvent,
+  onCollapseEvent,
+  cookieName,
+}: UseExpandedFrontpageSectionProps) => {
+  cookieName ??= `expand_frontpage_section_${section}`;
```

Comment:
This is a bit dangerous to do, because we're supposed to call `registerCookie()` in `cookies.ts` on every cookie that we add, and generating a default name here makes it easy for us to forget that step. I think we should just require a cookie name.

Reply:
Ah, good point. I wrote this before the cookie stuff reached master and missed this when merging.